File: typescript/repl/repl.ts
import { EOF } from "../char/index.ts";
import { IterableLexer, Lexer } from "../lexer/index.ts";
import { Parser } from "../parser/index.ts";

const PROMPT = ">> ";

async function prompt(prefix: string): Promise<string> {
  await Deno.stdout.write(new TextEncoder().encode(prefix));

  const buf = new Uint8Array(1024);
  const numberOfBytesRead = await Deno.stdin.read(buf);

  if (numberOfBytesRead === null) {
    return EOF;
  }

  const input = new TextDecoder().decode(buf.subarray(0, numberOfBytesRead))
    .trim();

  return input;
}

export async function repl(): Promise<void> {
  const encoder = new TextEncoder();

  while (true) {
    const line = await prompt(PROMPT);

    if (line === "exit") {
      break;
    }

    const lexer = new Lexer(line);
    const parser = new Parser(lexer);
    const program = parser.parseProgram();

    if (parser.errors.length > 0) {
      for (const error of parser.errors) {
        await Deno.stdout.write(encoder.encode(`${error}\n`));
      }
    }

    await Deno.stdout.write(encoder.encode(`${program.toString()}\n`));
  }
}

File: zig/src/repl/repl.zig
const std = @import("std");
const Lexer = @import("../lexer/lexer.zig").Lexer;
const TokenType = @import("../token/token_type.zig").TokenType;

pub fn repl(io: std.Io) !void {
    var writeBuffer: [1024]u8 = undefined;
    var stdoutWriter = std.Io.File.stdout().writer(io, &writeBuffer);
    const writer = &stdoutWriter.interface;
    const prefix = ">> ";

    while (true) {
        try writer.writeAll(prefix);
        try writer.flush();

        var readBuffer: [1024]u8 = undefined;
        var stdinReader = std.Io.File.stdin().reader(io, &readBuffer);
        const reader = &stdinReader.interface;
        const line = try reader.takeDelimiterExclusive('\n');

        if (std.mem.eql(u8, line, "exit")) {
            break;
        }

        var lexer = Lexer.init(line);
        var token = lexer.next_token();

        while (token != .EOF) : (token = lexer.next_token()) {
            var buf: [256]u8 = undefined;
            const str = try token.toString(&buf);

            try writer.print("{s}\n", .{str});
            try writer.flush();
        }
    }
}
